File: 198. House Robber/Approach03.py
from typing import *

class Solution:
    def rob(self, nums: List[int]) -> int:
        first_val = nums[0]
        second_val = 0
        
        for i in range(2, len(nums)+1):
            rob = nums[i-1] + second_val
            skip = first_val
            temp = max(rob, skip)
            second_val = first_val
            first_val = temp

        return first_val

File: 198. House Robber/Approach03.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();

        int first_val = nums[0];
        int second_val = 0;
        
        for(int i=2; i<=n; i++){
            int rob = nums[i-1] + second_val;
            int skip = first_val;
            int temp = max(rob, skip);
            second_val = first_val;
            first_val = temp;            
        }
        return first_val;
    }
};